File: src/easy/083-remove-duplicates-from-sorted-list/remove_duplicates_from_sorted_list.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        hashmap = {}

        current = head
        prev = ListNode(-1, current)

        while current:
            if current.val in hashmap:
                prev.next = current.next
            else:
                hashmap[current.val] = current.val
                prev = prev.next
            current = current.next

        return head

File: src/easy/083-remove-duplicates-from-sorted-list/removeDuplicatesFromSortedList.js
/**
 * Definition for singly-linked list.
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */

function ListNode(val, next) {
  this.val = (val===undefined ? 0 : val)
  this.next = (next===undefined ? null : next)
}

const { node1, node4, node9 } = require('./linkedLists');

var deleteDuplicates = function(head) {
  const map = {};

  let current = head;
  let prev = new ListNode(-1, current);

  while (current) {
    if (map[String(current.val)]) {
      prev.next = current.next;
    } else {
      map[current.val] = String(current.val);
      prev = prev.next;
    }
    current = current.next;
  }

  return head;
};

console.log(deleteDuplicates(node1));   // output = [1,2]
console.log(deleteDuplicates(node4));   // output = [1,2,3]
console.log(deleteDuplicates(node9));   // output = [0]
